### DISCUSSION TITLE
[Proxy Manager] Add the mapping feature

### DISCUSSION THREAD

**Discussion by frengooo** [mia-platform/community#33] [Ideas]
We need to proxy some APIs through the Mia-Platform environment of a client to an external server.
We thought to use the Mia-Platform `http-proxy-manager` plugin, but the client needs to call our proxy with a payload and to map it before proxy to the destination service.

e.g.
```
received payload:
{
  "name": "Valentino",
  "surname": "Rossi"
}

payload to be proxied:
{
  "firstName": "Valentino",
  "familyName": "Rossi"
}
```

What do you think about this new feature? Could we implement it?

**Comment by davidebianchi** [comment]
Hi @silversoul93, this could be an interesting feature to add, but has some important side effects like read all the body, save it in memory and convert it. Also, assume the content type of the body to be (or read the specific header, and in a first step support only JSON).

So, the flow which you need is:
* client send a request with a request body
* `http-proxy-manager` take the request and transform the body. The output should be totally defined? if there will be unknown fields, it will be forwarded (or set it as optional)?
* `http-proxy-manager` create the request and send it
* the service response to the `http-proxy-manager`
* `http-proxy-manager` respond to the client. The response should be transformed in some way?

**Reply by frengooo** [reply]
Hi @davidebianchi, thanks for your feedback.
We should support only JSON as first step by reading the specific header, for example like this:
```javascript
const contentTypeToMapper = {
  "application/json": mapJson,
}
...
const contentType = headers["content-type"]
const mapper = contentTypeToMapper[contentType]
if (!mapper) {
  reply.code(415).send({ msg: `Unsupported content type: ${contentType}` })
  return
}
```

The flow is right, following the replies to your concerns:
- `The output should be totally defined?` &rarr; I think we can handle this with a flag in the configuration, e.g. `isBodyMappingStrict`
- `The response should be transformed in some way?` &rarr; I think we don't need this feature too, but I will ask to the client to be sure

**Reply by frengooo** [reply]
@davidebianchi the client needs to transform the response too

**Reply by davidebianchi** [reply]
Is this feature planned? What is the priority about it?

**Reply by frengooo** [reply]
This feature is not planned yet. We are waiting for customer confirmation before planning it.
